crates/starlight/src/vm/promise.rs

```diff
@@ -0,0 +1,301 @@
+                vm,
+                &mut arguments,
+                JsValue::encode_undefined_value(),
+            );
+
+            match res {
+                Ok(_) => {}
+                Err(err) => {
+                    // should this reject the prom?
+                    println!(
+                        "prom func invoc failed: {}",
+                        err.to_string(vm).ok().expect("conversion failed")
+                    )
+                }
+            }
+
+            vm.remove_persistent_root(&prom_rooted_id);
+            vm.remove_persistent_root(&func_rooted_id);
+        });
+
+        sched_res.map(|_| promise)
+    }
+    pub fn new_unresolving(vm: &mut Runtime) -> Result<JsValue, JsValue> {
+        let proto = vm
+            .global_object()
+            .get(vm, "Promise".intern())?
+            .to_object(vm)?
+            .get(vm, "prototype".intern())?
+            .to_object(vm)?;
+
+        let structure = Structure::new_indexed(vm, Some(proto), false);
+        let mut obj = JsObject::new(vm, &structure, JsPromise::get_class(), ObjectTag::Ordinary);
+
+        *obj.data::<JsPromise>() = ManuallyDrop::new(JsPromise {
+            subs: vec![],
+            resolution: None,
+        });
+        Ok(JsValue::new(obj))
+    }
+    pub fn resolve(&mut self, vm: &mut Runtime, resolution: JsValue) -> Result<(), JsValue> {
+        self.do_resolve(vm, Ok(resolution))
+    }
+    pub fn reject(&mut self, vm: &mut Runtime, rejection: JsValue) -> Result<(), JsValue> {
+        self.do_resolve(vm, Err(rejection))
+    }
+    fn do_resolve(
+        &mut self,
+        vm: &mut Runtime,
+        resolution: Result<JsValue, JsValue>,
+    ) -> Result<(), JsValue> {
+        //println!("do_resolve, subs={}", self.subs.len());
+
+        if self.resolution.is_some() {
+            Err(JsValue::encode_object_value(JsString::new(
+                vm,
+                "Promise was already resolved",
+            )))
+        } else {
+            self.resolution = Some(resolution);
+
+            // todo everything below needs to be in async job.. need to root persistent again... later
+
+            if let Some(ok_resolution) = self.resolution.unwrap().ok() {
+                for sub in &self.subs {
+                    // invoke 0, resolve 3
+                    if let Some(jsFunc) = sub.0 {
+                        let this = JsValue::encode_undefined_value();
+                        let mut args_vec = vec![ok_resolution];
+                        let mut args = Arguments::new(this, args_vec.as_mut_slice());
+                        let sub_res = jsFunc
+                            .get_jsobject()
+                            .as_function_mut()
+                            .call(vm, &mut args, this);
+                        sub.3
+                            .get_jsobject()
+                            .as_promise_mut()
+                            .do_resolve(vm, sub_res)?
```

Comment:
When `do_resolve` or `do_reject`, the condition that `sub_res` is another promise should also be considered. if that, should "wait" the new promise `sub_res` to resolve to `do_resolve` sub.3.

Reply:
Here is a detailed spec https://promisesaplus.com/. I also implemented a js version before in here https://github.com/jameslahm/PromiseA.

Reply:
> When `do_resolve` or `do_reject`, the condition that `sub_res` is another promise should also be considered. if that, should "wait" the new promise `sub_res` to resolve to `do_resolve` sub.3.

Nice catch, thanks.. i'll add that to my todo list